### DISCUSSION TITLE
Find all simple cycles in directed graph

### DISCUSSION THREAD

**Discussion by tcardlab** [networkit/networkit#1036] [Q&A]
Is there a utility in NetworKit that can help me do this?

I know Graph-Tool and Networkx have functions to do this, i figured NetworKit might as well.

**Comment by fabratu** [comment]
Currently, there is no function which directly returns the found cycles. However, you can use `DFSFrom`[1] for every connected component of the graph (retrieved via the [components module](https://networkit.github.io/dev-docs/python_api/components.html)).

`DFSFrom` supports a callback function parameter, which lets you operate on the current node (and use graph/neighbor iterators). A minimal example would be:

```
import networkit as nk
G = nk.generators.ErdosRenyiGenerator(10, 0.3).generate()
result = []
nk.graph.Traversal.DFSfrom(G, source, lambda u: result.append(u))
```

[1] https://networkit.github.io/dev-docs/python_api/traversal.html#networkit.traversal.Traversal.DFSfrom
